[FLINK-20343] Add new Deployment overview documentation page

rmetzger opened the pull request:
For convenience, I rendered the page here:

![FireShot Capture 009 - Apache Flink 1 12-SNAPSHOT Documentation_ Clusters   Deployment - localhost](https://user-images.githubusercontent.com/89049/100239377-a1d62100-2f31-11eb-8392-812af25638b4.png)

XComp commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+
+
+<!-- Image source: https://docs.google.com/drawings/d/1EfloufuOp1A7YDwZmBEsHKRLIrrbtRkoWRPcfZI5RYQ/edit?usp=sharing -->
+<img width="100%" src="{% link fig/deployment_modes.svg %}" alt="Figure for Deployment Modes" />
```
Could we match the order of the job submission modes of the picture and the sections below?

tillrohrmann replied:
Maybe the marketing team could help us with beautifying this picture. No offense for your artistic skills Robert!

XComp commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+
+
+
+## Deployment Modes
```
We should name these things in the same way where possible: The overview calls this the "job submission modes".

XComp commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+If you don't know where to start, we recommend using the Command Line Interface for submitting Flink applications to a Standalone Cluster.
+
+<!-- Image source: https://docs.google.com/drawings/d/1s_ZlXXvADqxWfTMNRVwQeg7HZ3hN1Xb7goxDPjTEPrI/edit?usp=sharing -->
+<img width="100%" src="{% link fig/deployment_overview.svg %}" alt="Figure for Overview and Reference Architecture" />
```
That's personal taste now, but: Could we have a bit of margin around this picture here? :-)

rmetzger replied:
I'll look into it ;)

XComp reviewed (changes requested):
Thanks for the proposal. I have a few remarks which you might want to address. Good work. 👍

rmetzger commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+            <td>JobManager</td>
+            <td>
+                JobManager is the name of the central work coordination component of Flink. It has implementations for different resource providers, which differ on high-availability, resource allocation behavior and supported job submission modes. <br />
+                JobManager <a href="">modes for job submissions</a>:
+                <ul>
+                    <li><b>Session Mode</b>: one JobManager instance manages multiple jobs sharing the same cluster of TaskManagers</li>
+                    <li><b>Application Mode</b>: runs the cluster exclusively for one job. The job main method (or client) gets executed on the JobManager.</li>
+                    <li><b>Per-Job Mode</b>: runs the cluster exclusively for one job. The job main method (or client) runs only prior to the cluster creation.</li>
+                </ul>
```
I don't have a strong opinion on this one, as there is probably no right or wrong here.

I wanted to mention the three modes here already, because they are essential to understand for the resource provider detail pages.

rmetzger commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+<img width="100%" src="{% link fig/deployment_overview.svg %}" alt="Figure for Overview and Reference Architecture" />
+
+
+<table class="table table-bordered">
```
Yes, I'll fix that!

rmetzger commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+            </td>
+            <td>
+                <ul>
+                    <li><a href="">Kubernetes</a></li>
+                    <li><a href="">YARN</a></li>
+                    <li><a href="">Mesos</a></li>
+                    <li><a href="">Standalone</a> (this is the barebone mode that requires just JVMs to be launched. Deployment with Docker, Docker Swarm / Compose, non-native Kubernetes and other models is possible through manual setup in this mode)
+                    </li>
```
I agree my current order is messed up. I basically did: order by number of users + standalone mode ;).  I'll put standalone on top.

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+        <tr>
+            <td>Flink Client</td>
+            <td>
+              Flink batch or streaming applications are compiled into a dataflow graph, which is submitted to the JobManager.
```
Maybe
```suggestion
              Compiles batch or streaming applications into a dataflow graph, which it then submits to the JobManager.
```

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+            </td>
+            <td>
+                <ul>
+                    <li><a href="">Command Line Interface</a></li>
+                    <li><a href="">REST Endpoint</a></li>
+                    <li><a href="">SQL Client</a></li>
+                    <li><a href="">Python REPL</a></li>
+                    <li><a href="">Scala REPL</a></li>
```
links are missing.

rmetzger replied:
I wanted to wait for your restructuring PR to be merged before putting proper links.

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+                JobManager is the name of the central work coordination component of Flink. It has implementations for different resource providers, which differ on high-availability, resource allocation behavior and supported job submission modes. <br />
+                JobManager <a href="">modes for job submissions</a>:
+                <ul>
+                    <li><b>Session Mode</b>: one JobManager instance manages multiple jobs sharing the same cluster of TaskManagers</li>
+                    <li><b>Application Mode</b>: runs the cluster exclusively for one job. The job main method (or client) gets executed on the JobManager.</li>
+                    <li><b>Per-Job Mode</b>: runs the cluster exclusively for one job. The job main method (or client) runs only prior to the cluster creation.</li>
```
I'd suggest to put session mode last.

tillrohrmann replied:
```suggestion
                    <li><b>Per-Job Mode</b>: runs the cluster exclusively for one job. The job's main method (or client) runs only prior to the cluster creation.</li>
```

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+                JobManager <a href="">modes for job submissions</a>:
+                <ul>
+                    <li><b>Session Mode</b>: one JobManager instance manages multiple jobs sharing the same cluster of TaskManagers</li>
+                    <li><b>Application Mode</b>: runs the cluster exclusively for one job. The job main method (or client) gets executed on the JobManager.</li>
```
```suggestion
                    <li><b>Application Mode</b>: runs the cluster exclusively for one job. The job's main method (or client) gets executed on the JobManager.</li>
```

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+
+## Overview and Reference Architecture
+
+The figure below shows the building blocks of every Flink cluster. There is always somewhere a client running. It takes the code of the Flink applications, transforms it into a job graph and submits it to the JobManager.
```
Not sure about the backticks. If we decide to go this way, then it needs to be applied consistently throughout the whole document.

rmetzger replied:
How about writing it as JobGraph, so that the reader notices it as a Flink terminology. Adding Backticks everywhere could disrupt the reader's flow a bit.

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+        <tr>
+            <td>High Availability Service Provider</td>
+            <td>
+                Flink’s JobManager supports a high availability mode, where multiple JobManagers participate in a leader election process, resulting in an active JobManager and potentially multiple standby JobManagers, allowing for fast failover in case the active JobManager is lost.
```
Maybe: Flink's JobManager can be run in high availability mode which allows Flink to recover from JobManager faults. In order to failover faster, multiple standby JobManagers can be started to act as backups.

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+            <td>
+                Flink can be deployed through different Resource Provider Frameworks, such as Kubernetes, YARN or Mesos.
+            </td>
+            <td>See "JobManager" implementations above.</td>
```
Maybe set anchor link if possible.

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+## Deployment Modes
+
+Flink can execute applications in one of three ways:
+ - in Session Mode, 
+ - in a Per-Job Mode, or
+ - in Application Mode.
```
Should we order them 

1. application mode
2. per-job
3. application mode

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+submitted application. Applications executed in the same (session) cluster use, and consequently compete
+for, the same resources. This has the advantage that you do not pay the resource overhead of spinning up
+a full cluster for every submitted job. But, if one of the jobs misbehaves or brings down a Task Manager,
+then all jobs running on that Task Manager will be affected by the failure. This, apart from a negative
```
```suggestion
then all jobs running on that TaskManager (or task manager) will be affected by the failure. This, apart from a negative
```

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+*Session mode* assumes an already running cluster and uses the resources of that cluster to execute any 
+submitted application. Applications executed in the same (session) cluster use, and consequently compete
+for, the same resources. This has the advantage that you do not pay the resource overhead of spinning up
+a full cluster for every submitted job. But, if one of the jobs misbehaves or brings down a Task Manager,
```
Same here with "Task Manager"

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+then all jobs running on that Task Manager will be affected by the failure. This, apart from a negative
+impact on the job that caused the failure, implies a potential massive recovery process with all the 
+restarting jobs accessing the filesystem concurrently and making it unavailable to other services. 
+Additionally, having a single cluster running multiple jobs implies more load for the JobManager, who 
```
Here we write JobManager. Hence, I'd suggest to make it consistent.

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+the application finishes. With this architecture, the *Application Mode* provides the same resource isolation
+and load balancing guarantees as the *Per-Job* mode, but at the granularity of a whole application. Executing 
+the `main()` on the JobManager allows for saving the CPU cycles required, but also save the bandwidth required
+for downloading the dependencies locally. Furthermore, it allows for more even spread of the network load of
```
```suggestion
for downloading the dependencies locally. Furthermore, it allows for more even spread of the network load for
```

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+non-blocking, will lead to the "next" job starting before "this" job finishes.
+
+<div class="alert alert-info" markdown="span">
+  <strong>Attention:</strong> The Application Mode allows for multi-`execute()` applications but 
+  High-Availability is not supported in these cases. High-Availability in Application Mode is only
+  supported for single-`execute()` applications.
```
@kl0u are we failing if we are running a multi-execute job with HA enabled?

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+To provide a dependency, there are the following options:
+- files in the **`lib/` folder** are added to the classpath used to start Flink. It is suitable for libraries such as Hadoop or file systems not available as plugins. Beware that classes added here can potentially interfere with Flink, for example if you are adding a different version of a library already provided by Flink.
+
+- **`plugins/<name>/`** are loaded at runtime by Flink through separate classloaders to avoid conflicts with classes loaded and used by Flink. Only jar files which are prepared as [plugins]({% link ops/plugins.md %}) can be added here.
```
```suggestion
- **`plugins/<name>/`** are loaded at runtime by Flink through separate classloaders to avoid conflicts with classes loaded and used by Flink. Only jar files which are prepared as [plugins]({% link deployment/filesystems/plugins.md %}) can be added here.
```

tillrohrmann commented on docs/ops/deployment/overview.md:
```diff
@@ -0,0 +1,358 @@
+
+- **`plugins/<name>/`** are loaded at runtime by Flink through separate classloaders to avoid conflicts with classes loaded and used by Flink. Only jar files which are prepared as [plugins]({% link ops/plugins.md %}) can be added here.
+
+### Download Maven dependencies locally
```
Should this be part of the overview page? Maybe it is better suited for tips & tricks.

XComp replied:
I would agree that the whole Deployment Best Practices should be moved somewhere else. To me it feels like being out of scope for this page.

rmetzger replied:
I removed this part.

tillrohrmann reviewed (approved):
Thanks for creating this PR @rmetzger. I had a couple of comments. All in all this overview page is very good improvement over what we currently have. I'd be fine with going forward and merging it once the comments are resolved.

rmetzger commented:
Thank you for your feedback!
I addressed all your comments, except the concerns regarding the figures. Since the responsible person for graphics is out sick, I don't have much hope that we will have updated graphics available soon.
Would you be okay with merging the PR as is, and pushing the nicer images through a hotfix later?

tillrohrmann commented:
Sure, I think updating the figures can also happen later.

rmetzger commented:
Thanks. I noticed I forgot to update some more links. I'll merge the PR once this is addressed.
